File: MaiBaiApplication/TianShenCash/src/main/java/com/tianshen/cash/activity/LoginActivity.java
package com.tianshen.cash.activity;

import android.Manifest;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.telephony.TelephonyManager;
import android.text.Editable;
import android.text.TextUtils;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import com.baidu.location.BDLocation;
import com.tbruyelle.rxpermissions2.RxPermissions;
import com.tianshen.cash.R;
import com.tianshen.cash.base.BaseActivity;
import com.tianshen.cash.constant.GlobalParams;
import com.tianshen.cash.event.LoginSuccessEvent;
import com.tianshen.cash.model.TianShenLoginBean;
import com.tianshen.cash.model.User;
import com.tianshen.cash.net.api.SignIn;
import com.tianshen.cash.net.base.BaseNetCallBack;
import com.tianshen.cash.utils.LocationUtil;
import com.tianshen.cash.utils.LogUtil;
import com.tianshen.cash.utils.RegexUtil;
import com.tianshen.cash.utils.TianShenUserUtil;
import com.tianshen.cash.utils.ToastUtil;
import com.tianshen.cash.view.ChangeInterface;
import com.tianshen.cash.view.MyLoginEditText;
import com.umeng.analytics.MobclickAgent;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashSet;
import java.util.Set;

import cn.jpush.android.api.JPushInterface;
import io.reactivex.functions.Consumer;

public class LoginActivity extends BaseActivity implements View.OnClickListener {
    private final int LOGINSUCCESS = 5;
    private final int REGISTEREQUEST=6;
    private final int REGISTESUCCESS=7;
    private int mRegIdQueryTimes;

    private String mobile = "", password = "";
    private String mUniqueId;

    private ImageView img_back;
    private Button bt_login;
    private TextView tv_forget;

    private TextView tv_registe;
    private MyLoginEditText et_mobile, et_password;
    private SignIn mSignIn;

    private Handler mHandler;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        initData();
        mHandler = new Handler() {
            public void handleMessage(Message msg) {
                super.handleMessage(msg);
                if (380 == msg.what) {
                    String reg_id = JPushInterface.getRegistrationID(mContext);
                    LogUtil.d("ret", "reg_id = " + reg_id);
                    if (reg_id == null || "".equals(reg_id)) {
                        if (mRegIdQueryTimes == 7) {
                            login(mobile, password);
                            mRegIdQueryTimes = 0;
                            return;
                        }
                        Message msgNext = new Message();
                        msgNext.what = 380;
                        mHandler.sendMessageDelayed(msgNext, 500);
                        mRegIdQueryTimes++;
                        if (mRegIdQueryTimes == 1) {
                            ToastUtil
                                    .showToast(mContext, getResources().getString(R.string.initialization_please_wait));
                        }
                    } else {
                        login(mobile, password);
                        mRegIdQueryTimes = 0;
                    }
                }
            }
        };
    }

    public void initData() {

        RxPermissions rxPermissions = new RxPermissions(LoginActivity.this);
        rxPermissions.request(Manifest.permission.READ_PHONE_STATE).subscribe(new Consumer<Boolean>() {
            @Override
            public void accept(Boolean aBoolean) throws Exception {
                TelephonyManager TelephonyMgr = (TelephonyManager) getSystemService(TELEPHONY_SERVICE);
                mUniqueId = TelephonyMgr.getDeviceId();
            }
        });

        mSignIn = new SignIn(mContext);
    }

    @Override
    protected int setContentView() {
        return R.layout.activity_login;
    }

    @Override
    protected void findViews() {
        img_back = (ImageView) findViewById(R.id.img_back);
        bt_login = (Button) findViewById(R.id.bt_login);
        tv_forget = (TextView) findViewById(R.id.tv_forget);
        et_mobile = (MyLoginEditText) findViewById(R.id.et_mobile);
        et_password = (MyLoginEditText) findViewById(R.id.et_password);
        tv_registe=(TextView) findViewById(R.id.tv_registe);
    }

    @Override
    protected void setListensers() {
        img_back.setOnClickListener(this);
        bt_login.setOnClickListener(this);
        bt_login.setClickable(false);
        tv_forget.setOnClickListener(this);
        tv_registe.setOnClickListener(this);
        et_mobile.setChangeListener(new ChangeInterface() {
            @Override
            public void change(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void changeBefore(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void chageAfter(Editable s) {
                mobile = s.toString();
                if (mobile.length() == 11) {
                    if(RegexUtil.IsTelephone(mobile)){
                        if(!password.equals("")){
                            bt_login.setClickable(true);
                            bt_login.setBackgroundResource(R.drawable.select_bt);
                        }
                    }else{
                        ToastUtil.showToast(mContext,"不是合法的手机号码");
                    }
                } else {
                    bt_login.setClickable(false);
                    bt_login.setBackgroundResource(R.drawable.button_gray);
                }
            }
        });
        et_password.setChangeListener(new ChangeInterface() {
            @Override
            public void change(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void changeBefore(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void chageAfter(Editable s) {
                password = s.toString();
                if (!password.equals("") && mobile.length() == 11) {
                    if( RegexUtil.IsTelephone(mobile)) {
                        bt_login.setClickable(true);
                        bt_login.setBackgroundResource(R.drawable.select_bt);
                    }
                } else {
                    bt_login.setClickable(false);
                    bt_login.setBackgroundResource(R.drawable.button_gray);
                }
            }
        });
    }
    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.img_back:
                backActivity();
                break;
            case R.id.tv_forget:
                Bundle bundle=new Bundle();
                bundle.putString("type", GlobalParams.CHANGE_LOGIN_PASSWORD);
                gotoActivity(mContext,ForgetPasswordActivity.class,bundle);
                break;
            case R.id.bt_login:
                mobile = et_mobile.getText().toString().trim();
                password = et_password.getText().toString().trim();
                bt_login.setClickable(false);
                Message msg = new Message();
                msg.what = 380;
                mHandler.sendMessage(msg);
                break;
            case R.id.tv_registe:
                Intent intent=new Intent(mContext,RegisteActivity.class);
                startActivityForResult(intent,REGISTEREQUEST);
                overridePendingTransition(R.anim.push_right_in,R.anim.not_exit_push_left_out);
                break;
        }
    }

    public void login(final String mobile, final String password) {
        try {
            JSONObject json = new JSONObject();
            json.put("device_id", mUniqueId);
            json.put("mobile", mobile);
            json.put("password", password);

            String jpushId = TianShenUserUtil.getUserJPushId(mContext);
            if (TextUtils.isEmpty(jpushId)) {
                jpushId = JPushInterface.getRegistrationID(mContext);
            }
            json.put("push_id", jpushId);
            final String finalJpushId = jpushId;
            mSignIn.signIn(json, bt_login, true, new BaseNetCallBack<TianShenLoginBean>() {
                @Override
                public void onSuccess(TianShenLoginBean paramT) {
//                    SharedPreferencesUtil.getInstance(mContext).putString(GlobalParams.JPUSH_ID_KEY, finalPush_id);
//                    UserUtil.setLoginPassword(mContext, password);
//                    setResult(LOGINSUCCESS);
//                    Set<String> tags=new HashSet<String>();
//                    BDLocation bdLocation = LocationUtil.getInstance(mContext).getLocation();
//                    if (bdLocation != null) {
//                        tags.add(bdLocation.getCityCode());
//                        tags.add(bdLocation.getCountryCode());
//                        tags.add(bdLocation.getProvince());
//                        JPushInterface.setAliasAndTags(mContext, TianShenUserUtil.getUserId(mContext), tags);
//                    }
//                    new SendBroadCastUtil(mContext).sendBroad(GlobalParams.REFRESH_HOME_PAGE_ACTION,null);
//                    new SendBroadCastUtil(mContext).sendBroad(GlobalParams.LOGIN_SUCCESS_ACTION,null);
//                    //Todo
//                    backActivity();

                    //保存用户信息
                    TianShenUserUtil.saveUserToken(mContext, paramT.getData().getToken());
                    TianShenUserUtil.saveUserId(mContext, paramT.getData().getCustomer_id());
                    TianShenUserUtil.saveUserPhoneNum(mContext, mobile);
                    TianShenUserUtil.saveUserJPushId(mContext, finalJpushId);


                    Set<String> tags = new HashSet<>();
                    BDLocation bdLocation = LocationUtil.getInstance(mContext).getLocation();
                    if (bdLocation != null) {
                        tags.add(bdLocation.getCityCode());
                        tags.add(bdLocation.getCountryCode());
                        tags.add(bdLocation.getProvince());
                        JPushInterface.setAliasAndTags(mContext, TianShenUserUtil.getUserId(mContext), tags);
                    }

                    gotoActivity(mContext, MainActivity.class, null);
                    EventBus.getDefault().post(new LoginSuccessEvent());

                }
                @Override
                public void onFailure(String url, int errorType, int errorCode) {
                    bt_login.setClickable(true);
                }
            });
        } catch (JSONException e) {
            e.printStackTrace();
            MobclickAgent.reportError(mContext, LogUtil.getException(e));
        }
    }

    @Override
    protected void onActivityResult(int arg0, int arg1, Intent arg2) {
        super.onActivityResult(arg0, arg1, arg2);
        switch (arg0){
            case REGISTEREQUEST:
                if(arg1==REGISTESUCCESS){
                    backActivity();
                }
                break;
        }
    }


    /**
     * 收到了在注册页面登录成功的消息
     */
    @Subscribe
    public void onLoginSuccess(LoginSuccessEvent event) {
        backActivity();
    }
}

File: MaiBaiApplication/TianShenCash/src/main/java/com/tianshen/cash/activity/LoginActivity.kt
package com.tianshen.cash.activity

import android.Manifest
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.telephony.TelephonyManager
import android.text.*
import android.text.method.LinkMovementMethod
import android.text.style.BackgroundColorSpan
import android.text.style.ClickableSpan
import android.text.style.ForegroundColorSpan
import android.view.View
import android.view.WindowManager
import cn.jpush.android.api.JPushInterface
import com.tbruyelle.rxpermissions2.RxPermissions
import com.tianshen.cash.R
import com.tianshen.cash.base.BaseActivity
import com.tianshen.cash.base.MyApplicationLike
import com.tianshen.cash.constant.GlobalParams
import com.tianshen.cash.event.LoginSuccessEvent
import com.tianshen.cash.model.CashAmountBean
import com.tianshen.cash.model.TianShenLoginBean
import com.tianshen.cash.net.api.GetCashAmountService
import com.tianshen.cash.net.api.SignIn
import com.tianshen.cash.net.base.BaseNetCallBack
import com.tianshen.cash.net.base.UserUtil
import com.tianshen.cash.utils.*
import com.umeng.analytics.MobclickAgent
import io.reactivex.functions.Consumer
import kotlinx.android.synthetic.main.activity_login2.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.json.JSONException
import org.json.JSONObject
import java.util.*

class LoginActivity : BaseActivity() {
    override fun setContentView(): Int = R.layout.activity_login2
    var str = "客服电话: 400-000-8685"
    var phoneNum = "4000008685"
    var hasTelephonePermission: Boolean? = false
    var mUniqueId: String? = null
    var mobile: String? = null
    var password: String? = null
    var mHandler: Handler? = null
    var mRegIdQueryTimes = 0
    var mSignIn: SignIn? = null
    override fun findViews() {
        initSpanString();
    }

    private fun initSpanString() {
        val ss = SpannableStringBuilder(str)
        ss.setSpan(object : ClickableSpan() {
            override fun onClick(widget: View) {
                callPhone()
            }

            override fun updateDrawState(ds: TextPaint?) {
//                super.updateDrawState(ds)
                ds?.isUnderlineText = false;
            }
        }, 6, str.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
        //span 字体颜色
        ss.setSpan(ForegroundColorSpan(resources.getColor(R.color.global_popular_color)), 6, str.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
        // span 默认背景颜色
        ss.setSpan(BackgroundColorSpan(Color.TRANSPARENT), 6, str.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
        tv_service_phone.highlightColor = Color.TRANSPARENT
        tv_service_phone.setText(ss);
        tv_service_phone.movementMethod = LinkMovementMethod.getInstance()
    }

    //打电话
    private fun callPhone() {
        val rxPermissions = RxPermissions(this)
        rxPermissions.request(Manifest.permission.CALL_PHONE).subscribe(Consumer<Boolean> {
            if (it) {
                val intent = Intent(Intent.ACTION_CALL, Uri.parse("tel:" + phoneNum))
                startActivity(intent)
            }
        })

    }

    override fun setListensers() {
        in_back.setOnClickListener { backActivity() }
        tv_login.setOnClickListener { loginHandle() }
        tv_forget_pwd.setOnClickListener { forgetPwd() }
        tv_regist.setOnClickListener { gotoActivity(mContext, RegisteActivity::class.java, null) }
        initData()
        initHandler()
    }

    //登录
    private fun loginHandle() {
        mobile = et_phone_number.text.toString().trim()
        password = et_login_pwd.text.toString().trim()
        //检查账号
        if (!TextUtils.isEmpty(mobile) and RegexUtil.IsTelephone(mobile)) {
            if (TextUtils.isEmpty(password)) {
                return;
            } else {
                tv_login.isClickable = false
                val msg = Message()
                msg.what = 380
                mHandler?.sendMessage(msg)
            }
        } else {
            ToastUtil.showToast(mContext, "不是合法的手机号码")
            return
        }

    }

    private fun initHandler() {
        mHandler = object : Handler() {
            override fun handleMessage(msg: Message) {
                super.handleMessage(msg)
                if (380 == msg.what) {
                    val app = MyApplicationLike.getMyApplicationLike()
                    val reg_id = JPushInterface.getRegistrationID(app)
                    LogUtil.d("ret", "reg_id = " + reg_id)
                    if (TextUtils.isEmpty(reg_id)) {
                        if (mRegIdQueryTimes == 7) {
                            login(mobile, password)
                            mRegIdQueryTimes = 0
                            return
                        }
                        val msgNext = Message()
                        msgNext.what = 380
                        mHandler?.sendMessageDelayed(msgNext, 500)
                        mRegIdQueryTimes++
                        if (mRegIdQueryTimes == 1) {
                            ToastUtil.showToast(mContext, resources.getString(R.string.initialization_please_wait))
                        }
                    } else {
                        login(mobile, password)
                        mRegIdQueryTimes = 0
                    }
                }
            }
        }
    }


    fun initData() {

        val rxPermissions = RxPermissions(this@LoginActivity)
        rxPermissions.request(Manifest.permission.READ_PHONE_STATE).subscribe { aBoolean ->
            if (aBoolean) {
                val TelephonyMgr = getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
                mUniqueId = TelephonyMgr.deviceId
                UserUtil.setDeviceId(mContext, TelephonyMgr.deviceId)
            } else {
                ToastUtil.showToast(mContext, resources.getString(R.string.please_open_permission))
            }
            hasTelephonePermission = aBoolean
        }

        mSignIn = SignIn(mContext)
    }

    fun login(mobile: String?, password: String?) {
        if (hasTelephonePermission == false) {
            ToastUtil.showToast(mContext, resources.getString(R.string.please_open_permission))
            return
        }
        if (hasTelephonePermission == true && TextUtils.isEmpty(mUniqueId)) {
            mUniqueId = System.currentTimeMillis().toString()
        }

        val app = MyApplicationLike.getMyApplicationLike()

        try {
            val json = JSONObject()
            json.put("device_id", mUniqueId)
            json.put("mobile", mobile)
            json.put("password", password)

            var jpushId = TianShenUserUtil.getUserJPushId(mContext)
            if (TextUtils.isEmpty(jpushId)) {
                jpushId = JPushInterface.getRegistrationID(app)
            }
            json.put("push_id", jpushId)
            val finalJpushId = jpushId
            mSignIn?.signIn(json, tv_login, true, object : BaseNetCallBack<TianShenLoginBean> {
                override fun onSuccess(paramT: TianShenLoginBean) {
                    //保存用户信息
                    TianShenUserUtil.saveUserToken(mContext, paramT.data.token)
                    TianShenUserUtil.saveUserId(mContext, paramT.data.customer_id)
                    TianShenUserUtil.saveUserPhoneNum(mContext, mobile)
                    TianShenUserUtil.saveUserJPushId(mContext, finalJpushId)


                    val tags = HashSet<String>()
                    val bdLocation = LocationUtil.getInstance().location
                    if (bdLocation != null) {
                        tags.add(bdLocation.cityCode)
                        tags.add(bdLocation.countryCode)
                        tags.add(bdLocation.province)
                        JPushInterface.setAliasAndTags(app, TianShenUserUtil.getUserId(app), tags)
                        if (JPushInterface.isPushStopped(app)) {
                            JPushInterface.resumePush(app)
                        }
                    }
                    checkStatusGoActivity()
                }

                override fun onFailure(url: String, errorType: Int, errorCode: Int) {
                    tv_login.isClickable = true
                }
            })
        } catch (e: JSONException) {
            e.printStackTrace()
            MobclickAgent.reportError(mContext, LogUtil.getException(e))
        }

    }

    private fun forgetPwd() {
        val bundle = Bundle()
        bundle.putString("type", GlobalParams.CHANGE_LOGIN_PASSWORD)
        gotoActivity(mContext, ForgetPasswordActivity::class.java, bundle)
    }

    /**
     * 根据当前用户的状态跳转到不同的页面
     */
    private fun checkStatusGoActivity() {

        val jsonObject = JSONObject()
        try {
            jsonObject.put(GlobalParams.USER_CUSTOMER_ID, TianShenUserUtil.getUserId(mContext))

        } catch (e: JSONException) {
            e.printStackTrace()
        }

        val getCashAmountService = GetCashAmountService(mContext)
        getCashAmountService.getData(jsonObject, object : BaseNetCallBack<CashAmountBean> {
            override fun onSuccess(bean: CashAmountBean) {

                val is_payway = bean.data.is_payway
                val cash_amount_status = bean.data.cash_amount_status

                val cur_credit_step = bean.data.cur_credit_step
                val total_credit_step = bean.data.total_credit_step
                val cash_amount = bean.data.cash_amount
                val show_dialog = bean.data.show_dialog

                var totalCreditStep = 0
                var curCreditStep = 0
                var cashAmount = 0

                if (!TextUtils.isEmpty(cur_credit_step)) {
                    curCreditStep = Integer.parseInt(cur_credit_step)
                }
                if (!TextUtils.isEmpty(total_credit_step)) {
                    totalCreditStep = Integer.parseInt(total_credit_step)
                }
                if (!TextUtils.isEmpty(cash_amount)) {
                    cashAmount = Integer.parseInt(cash_amount)
                }
                if ("0" == show_dialog) { //掌众认证失败跳转到首页
                    gotoActivity(mContext, MainActivity::class.java, null)
                    EventBus.getDefault().post(LoginSuccessEvent())
                    return
                }
                if (curCreditStep > 0 && curCreditStep < totalCreditStep) { //跳转到认证中心页面
                    gotoActivity(mContext, AuthCenterMenuActivity::class.java, null)
                    EventBus.getDefault().post(LoginSuccessEvent())
                    finish()
                    return
                }
                if ("0" == cash_amount_status) { //跳转到首页
                    if (cashAmount == 0 && curCreditStep > 0) { //跳转到认证中心页面
                        gotoActivity(mContext, AuthCenterMenuActivity::class.java, null)
                    } else {//跳转到首页
                        gotoActivity(mContext, MainActivity::class.java, null)
                    }
                    EventBus.getDefault().post(LoginSuccessEvent())
                } else if ("1" == cash_amount_status && "0" == is_payway) { //跳转到首页
                    gotoActivity(mContext, MainActivity::class.java, null)
                    EventBus.getDefault().post(LoginSuccessEvent())
                } else if ("1" == cash_amount_status && "1" == is_payway) { //跳转到掌众借款页面
                    gotoActivity(mContext, ConfirmBorrowingActivity::class.java, null)
                    finish()
                } else if ("2" == cash_amount_status) { //跳转到跑分等待页面
                    gotoActivity(mContext, EvaluateAmountActivity::class.java, null)
                    finish()
                }
            }

            override fun onFailure(url: String, errorType: Int, errorCode: Int) {

            }
        })
    }

    /**
     * 收到了在注册页面登录成功的消息
     */
    @Subscribe
    fun onLoginSuccess(event: LoginSuccessEvent) {
        backActivity()
    }
}
